pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -69,51 +71,93 @@ func (s *StreamWriter) Close() error {
 
 // StreamReader is a struct for reading piece download stream from server
 type StreamReader struct {
-	stream    pb.PieceStoreRoutes_RetrieveClient
-	src       *utils.ReaderSource
-	totalRead int64
+	throttle     *sync2.Throttle
+	stream       pb.PieceStoreRoutes_RetrieveClient
+	src          *utils.ReaderSource
+	totalRead    int64
+	MaxAllocSize int64
+	allocated    int64
 }
 
 // NewStreamReader creates a StreamReader for reading data from the piece store server
-func NewStreamReader(signer *Client, stream pb.PieceStoreRoutes_RetrieveClient, pba *pb.PayerBandwidthAllocation) *StreamReader {
+func NewStreamReader(signer *Client, stream pb.PieceStoreRoutes_RetrieveClient, pba *pb.PayerBandwidthAllocation, MaxAllocSize int64) *StreamReader {
 	sr := &StreamReader{
-		stream: stream,
+		throttle:     sync2.NewThrottle(),
+		stream:       stream,
+		MaxAllocSize: MaxAllocSize,
 	}
 
-	sr.src = utils.NewReaderSource(func() ([]byte, error) {
-		updatedAllocation := int64(signer.bandwidthMsgSize) + sr.totalRead
-		allocationData := &pb.RenterBandwidthAllocation_Data{
-			PayerAllocation: pba,
-			Total:           updatedAllocation,
-		}
-
-		serializedAllocation, err := proto.Marshal(allocationData)
-		if err != nil {
-			return nil, err
+	// Send signed allocations to the piece store server
+	go func() {
+		trustedSize := int64(signer.bandwidthMsgSize)
+
+		// Allocate until we've reached the file size
+		for sr.allocated < MaxAllocSize {
+
+			nextAllocSize := trustedSize
+			if sr.allocated+trustedSize > MaxAllocSize {
+				nextAllocSize = MaxAllocSize - sr.allocated
+			}
+
+			allocationData := &pb.RenterBandwidthAllocation_Data{
+				PayerAllocation: pba,
+				Total:           sr.allocated + nextAllocSize,
+			}
+
+			serializedAllocation, err := proto.Marshal(allocationData)
+			if err != nil {
+				sr.throttle.Fail(err)
+				return
+			}
+
+			sig, err := signer.sign(serializedAllocation)
+			if err != nil {
+				sr.throttle.Fail(err)
+				return
+			}
+
+			msg := &pb.PieceRetrieval{
+				Bandwidthallocation: &pb.RenterBandwidthAllocation{
+					Signature: sig,
+					Data:      serializedAllocation,
+				},
+			}
+
+			if err = stream.Send(msg); err != nil {
+				sr.throttle.Fail(err)
+				return
+			}
+
+			sr.allocated += trustedSize
+
+			if err = sr.throttle.ProduceAndWaitUntilBelow(nextAllocSize, MaxAllocSize*2); err != nil {
```

Comment:
`int64(signer.bandwidthMsgSize*2)` or `*3`... the ideal would such number that the next allocation reaches the same time the storagenode runs out of allocations (with some margin of error)

Reply:
Are you referring to int64(signer.bandwidthMsgSize*2) instead of nextAllocSize?

Reply:
`sr.throttle.ProduceAndWaitUntilBelow(nextAllocSize, int64(signer.bandwidthMsgSize*2))`

Reply:
this assumes that the storage node is always sending back in chunks of `bandwidthMsgSize`